src/Symfony/Component/Security/Core/Role/RoleHierarchy.php

```diff
@@ -76,6 +74,66 @@ protected function buildRoleMap()
             }
 
             $this->map[$main] = array_unique($this->map[$main]);
+
+            if (str_contains($main, '*') && false !== ($pattern = $this->getPlaceholderPattern($main))) {
+                $this->rolePlaceholdersPatterns[$main] = $pattern;
+            }
         }
     }
+
+    private function resolveReachableRoleNames(array $roles, array &$visitedPlaceholders = []): array
+    {
+        $reachableRoles = $roles;
+
+        foreach ($roles as $role) {
+            if (!isset($this->map[$role])) {
+                continue;
+            }
+
+            foreach ($this->map[$role] as $r) {
+                $reachableRoles[] = $r;
+            }
+        }
+
+        $placeholderRoles = array_diff($this->getMatchingPlaceholders($reachableRoles), $visitedPlaceholders);
+        if (!empty($placeholderRoles)) {
+            array_push($visitedPlaceholders, ...$placeholderRoles);
+            $resolvedPlaceholderRoles = $this->resolveReachableRoleNames($placeholderRoles, $visitedPlaceholders);
+            foreach (array_diff($resolvedPlaceholderRoles, $placeholderRoles) as $r) {
+                $reachableRoles[] = $r;
+            }
+        }
+
+        return $reachableRoles;
+    }
+
+    private function getMatchingPlaceholders(array $roles): array
+    {
+        $resolved = [];
+
+        foreach ($this->rolePlaceholdersPatterns as $placeholder => $pattern) {
+            if (!\in_array($placeholder, $resolved) && \count(preg_grep($pattern, $roles) ?? null)) {
+                $resolved[] = $placeholder;
+            }
+        }
+
+        return $resolved;
+    }
+
+    /**
+     * Build the regex pattern for the given role:
+     *   - Replace valid wildcards with a non-wildcard matching pattern and
+     *   - Escape reserved regex characters.
+     *
+     * A valid wildcard is a * prefixed with _ and immediately followed by _ or EOL.
+     *
+     * @return string|false The regex pattern, or false if there is no valid wildcard in the role
+     */
+    private function getPlaceholderPattern(string $role): string|false
+    {
+        /** @var int $count */
+        $placeholderPattern = preg_replace(pattern: '/(?<=_)\\\\\*(?=_|$)/', replacement: '[^\*]*', subject: preg_quote($role), count: $count);
+
+        return ($count > 0) ? sprintf('/%s/', $placeholderPattern) : false;
```

Comment:
⚠️  Security risk there

```php
        return ($count > 0) ? sprintf('/%s/', $placeholderPattern) : false;
```

If you use only `/%s/`, it's as if you wrote "`*ROLE_*`" .. i think `/^%s/` would be better no ?

I suggest adding this test: 

````
$role = new RoleHierarchy(['ROLE_FOO_*' => ['ROLE_FOO_BAR']]));
$this->assertEquals(['ROLE_ROLE_FOO_BAR'], $role->getReachableRoleNames(['ROLE_ROLE_FOO_BAR'])); 
````

Reply:
Good catch, thanks! I'd also add a trailing `$` to strictly match placeholders like `ROLE_FOO_*_BAR`

Reply:
Should be fixed with [1db2351](https://github.com/symfony/symfony/pull/52099/commits/1db235197e53a1351a14200fd1b7245983d71e40)